### DISCUSSION TITLE
API application client_credential error

### DISCUSSION THREAD

**Discussion by klinux** [zitadel/zitadel#7426] [Q&A] [answered]
Hi,

I'm trying to create a API application and create a token, but I'm receiving this error:

```
{
    "error": "server_error",
    "error_description": "ID=QUERY-Dfbg2 Message=Errors.User.NotFound Parent=(sql: no rows in result set)"
}
```
Here the example of my request:

```
curl --location 'https://<HOST>/oauth/v2/token' \
--header 'Content-Type: application/x-www-form-urlencoded' \
--header 'Authorization: Basic <CLIENT_ID+CLIENT_SECRET>' \
--data 'grant_type=client_credentials&scope=openid'
```
When I use a **Service User**, works fine, but with **API application** doesn't.

Thanks

**Comment by dakshitha** [comment, marked as answer]
It seems like there might be a misunderstanding regarding the use of an API application's Client ID and Client Secret for token acquisition. In the context of ZITADEL and OpenID Connect (OIDC), the process you've described is not applicable for API application types. 

1. Tokens are typically issued in the context of a user. This means:
   - For **human users**, authentication must happen through one of the standard OIDC login flows (such as web application, native application, or user agent flows). These flows are designed to handle user interactions.
   - For **machine-to-machine interactions** (where there's no human user involved), a "service user" must be created. Tokens for service users are obtained using the client credentials grant type (or PAT), which is why your API call succeeded in this context.

2. The Client ID and Client Secret of an API application are intended for different purposes:
   - To authenticate the API application itself to ZITADEL for operations like token introspection. 
   - Unlike service users, API applications' credentials cannot be used to directly generate tokens via the token endpoint. Their role is more about managing and validating tokens rather than obtaining them.

Could you please elaborate your use case? This information will help us guide you better.

**Reply by klinux** [reply]
Thank you @dakshitha I'm user from Keycloak, and I really misunderstanding that.

**Reply by dakshitha** [reply]
You're welcome! :)
